### ISSUE TITLE
libcxx update

### ISSUE THREAD

**Issue by glebm** [OpenOrbis/OpenOrbis-PS4-Toolchain#262]
It'd be great if libcxx could be updated to a more recent upstream. The latest release is 3 years old at this point and it'd be great to have some of the new features (especially for cross-platform projects like DevilutionX, https://github.com/diasurgical/devilutionX/pull/6929).

Posting the feature request here because https://github.com/OpenOrbis/llvm-project does not allow posting Issues.

**Comment by seuros** [comment]
@glebm , do you want to pair on this ?  I have some progress locally.

**Comment by glebm** [comment]
@seuros I don't know much about PS4 (I'm not the author of DevilutionX PS4 port), so I'm not sure if I'll be of much help here.
I do know C++ fairly well.

**Comment by kiwidoggie** [comment]
If you are willing to help, we are in process of moving to a completely new version of the toolchain using BSD libc + latest LLVM. It's currently a private project with many smaller things needed to be completed

**Comment by glebm** [comment]
I'm happy to test DevilutionX with it on CI once a Docker image is available. :+1: 
DevilutionX is a C++17 codebase (using a bit of C++20 where available), if it works then it is likely a lot of other things will do too.

**Comment by glebm** [comment]
2 years down the road, DevilutionX is now a C++23 codebase, so a new toolchain is needed if we are to keep PS4 support (e.g. https://github.com/diasurgical/DevilutionX/pull/8615).

**Comment by kiwidoggie** [comment]
One is in development, we are looking for help :)

**Comment by glebm** [comment]
Where can I follow the work? Is it still a private project?

**Comment by seuros** [comment]
I will try to compile Devilution this weekend fir the PS4

**Comment by glebm** [comment]
I've also found the Discord and found the following repo mentioned on there: https://github.com/xinitrcn1/scarytoolchain-ps4 by @xinitrcn1.
It looks like an llvm20 toolchain, implemented as a bunch of patch files. It's musl, not BSD libc, so I'm guesing it's not the one that @kiwidoggie is working on.

**Comment by seuros** [comment]
The toolchain still not complete and still private in this org, i had started the update of llvm https://github.com/seuros/llvm-project/tree/openorbis/20.x.

I will find a slot this weekend and rebase/test this project.  A lot of things changed since then.

**Comment by kiwidoggie** [comment]
> Where can I follow the work? Is it still a private project?

Yes but if interested join the discord and I can get you access to everything :)

At this point I rather get everything near perfect, with compiling a new toolchain, headers, stubs, samples, libraries from the ground up (most of this work is near 90% complete already)

TODO: insert online documentation link here

@glebm 

The new toolchain uses bsd libc, with latest llvm toolchain (21?). The lld fork will need major fixes as when going through it with a C++ port of create-fself (which will also be OS'd) there was a metric fuckton wrong with the original and the lld port. The updated create-fself-cpp should be 100% accurate for both libraries and eboots.

@seuros I'd hold off on updating/rebasing as it's going to need a major refactoring anyway.

**Comment by xinitrcn1** [comment]
i made it so it can easily be upgraded to latest llvm from master
my idea was to keep it with patches and manually upgrade to master branches as time went on
i've already been working on a rework for the whole thing which i should push now

and yes i just pushed, you can look at the fact ive been working on it for... like 30 seconds, and it just werks
notably fstat isn't fixed yet but pthreads and recursive mutexes are

all in all **im forced** to use this scary toolchain because i have projects
which use C++23 stuff which llvm19 is too scared to have, so i have to use llvm20+ or greater

i should note the scripts are for freeBSD only, and nobody is able to reproduce my prefix
which is why I usually just send the whole prefix in tar balls to interested people
is it a good method? not at all - but that's the state of "i wanna use llvm21 on orbis"
